lib/pure/pegs.nim

```diff
@@ -85,6 +85,11 @@ type
     of pkBackRef..pkBackRefIgnoreStyle: index: range[0..MaxSubpatterns]
     else: sons: seq[Peg]
   NonTerminal* = ref NonTerminalObj
+  NtCallback* = object
+    enter*: proc(nt: NonTerminal, matchStart: int)
+    leave*: proc(nt: NonTerminal; matchStart, matchLength: int)
+  Callbacks* = ref object
```

Comment:
This indirection via ``ref`` is not necessary. Default can be an object with ``nil`` callbacks or callbacks that contain a ``discard`` body.

Reply:
In fact, this ``Callbacks`` type shouldn't exist.

Reply:
Your'e right. I thought maybe ``Callbacks`` could get more fields later, introducing callbacks for other ``PegKind``s, but that would break the API anyway.

Reply:
Not if you add a constructor for ``NtCallback`` and you should.

Reply:
Sry for not getting it: adding fields to ``Callbacks`` later would break backwards compatibility, how would a constructor for ``NtCallback`` help me with that?

Reply:
The constructor can be adapted to give default values to the newly introduced fields.

Reply:
Yes, but that would have to be a constructor for ``Callbacks``, not for ``NtCallbacks``, because the new fields would be added to ``Callbacks``. They would be of a type similar to ``NtCallback``, in that they would have ``enter`` and ``leave`` fields for the actual callback procs, but the signatures of these procs would be different than those in ``NtCallback``.

Are you maybe thinking of a "flat" callback container object like this:
```
type WhatEverTheNameIs = object
  enterNonTerminal: ...
  leaveNonTerminal: ...
  enterOrderedChoice: ...
  leaveOrderedChoice: ...
```

Reply:
> Are you maybe thinking of a "flat" callback container object like this:

Yes, now we are on the same page.

Reply:
> Yes, now we are on the same page.

Nice, will make it so.

---------- SNIP ----------------
Please just ignore the following if you've got enough of the discussion. :o)

How about a case object type as an alternative:
```
type PegCallbacks = object
  case kind: PegKind
  of pkNonTerminal:
    enterNonTerminal: <specific enter callback's signature>
    leaveNonTerminal: <specific leave callback's signature>
  of pkOrderedChoice:
    enterOrderedChoice: <specific enter callback's signature>
    leaveOrderedChoice: <specific leave callback's signature>
...
  else:
    discard
```
The ``parse`` proc could accept a set of these, unique as per ``kind`` value.

Reply:
For the do-nothing constructor, ``PegCallbacks()``, we would just have to add a new ``PegKind`` ``pkNone`` at the top.

Reply:
The do-nothing default for calling ``rawMatch`` without callbacks would just be an empty set of ``PegCallbacks`` objects.

Reply:
Oh, and the ``else`` branch in the case object is probably a bad idea to begin with...